### ISSUE TITLE
Pyright make false assumption about size of tuple

### ISSUE THREAD

**Issue by CyberDem0n** [microsoft/pyright#5968]
**Code or Screenshots**

```python
from typing import Any


def test(action: str, *args: Any) -> None:
    if action == 'create' or len(args) == 1:
        print(args[0])

    if action == 'create' and len(args) > 1:
        print(args[1])


test('create', 1, 2)
```

```bash
$ pyright testpyright.py
testpyright.py
  testpyright.py:9:15 - error: Index 1 is out of range for type tuple[Any] (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```

**Which version?**

1.1.327

Previous version didn't had this problem.

**Comment by erictraut** [comment]
Thanks for the bug report.

This is an interesting case. The initial type of `args` is `tuple[Any, ...]`. This is shorthand for a union of all tuples that contain `Any` entries regardless of length, so it's equivalent to `tuple[()] | tuple[Any] | tuple[Any, Any] | ...`. When pyright determines whether the subscript is within range when indexing into the tuple, it ignores the fact that it's theoretically out of range for some subtypes in the union. That's why `args[0]` doesn't generate any error condition.

The `len(args) == 1` expression is a supported type guard, and it narrows the type of `args` from `tuple[Any, ...]` to `tuple[Any]`.

The type of `args` after the first `if` block becomes `tuple[Any, ...] | tuple[Any]`, which should be treated the same as just `tuple[Any, ...]` when considering whether a subscript is out of range, but pyright was considering both subtypes independently of each other, and it therefore emitted an error when attempting to access index `1` on a value of type `tuple[Any]`.

I've updated the logic to handle this case. It now considers whether a tuple of known length is subsumed by a tuple of unbounded length before emitting an error about an out-of-bound index. This prevents the false positive error in your code sample. This fix will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.328, which I just published. It will also be included in a future release of pylance.

**Comment by CyberDem0n** [comment]
Hi @erictraut, there is a regression. Starting from 1.1.348 it again falsely reports `Index 1 is out of range` :(

**Comment by erictraut** [comment]
@CyberDemOn, this was an intentional change due to a clarification in the typing spec about how `tuple[Any, ...]` works. Pyright's current behavior is "as designed", so I don't consider it a bug.
